packages/dds/tree/src/core/undo/undoRedoManager.ts

```diff
@@ -0,0 +1,122 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ChangeFamily, ChangeFamilyEditor } from "../change-family";
+import { GraphCommit, GraphCommitType, RevisionTag, tagChange } from "../rebase";
+import { ReadonlyRepairDataStore, RepairDataStore } from "../repair";
+import { Delta } from "../tree";
+
+/**
+ * Manages the undo commit tree and repair data associated with undoable commits.
+ */
+export class UndoRedoManager<TChange, TEditor extends ChangeFamilyEditor> {
+	private pendingRepairData?: RepairDataStore;
+
+	/**
+	 * @param repairDataStoryFactory - Factory function for creating {@link RepairDataStore}s to create and store repair
+	 * data for {@link UndoableCommit}s.
+	 * @param changeFamily - {@link ChangeFamily} used for inverting changes.
+	 * @param applyChange - Callback to apply undos as local changes. This should call {@link UndoRedoManager.trackCommit}
+	 * with the created commit.
+	 * @param headUndoCommit - Optional commit to set as the initial undoable commit.
+	 */
+	public constructor(
+		private readonly repairDataStoryFactory: () => RepairDataStore,
+		private readonly changeFamily: ChangeFamily<TEditor, TChange>,
+		protected readonly applyChange: (
+			change: TChange,
+			revision?: RevisionTag,
+			type?: GraphCommitType,
+		) => void,
+		protected headUndoableCommit?: UndoableCommit<TChange>,
+	) {}
+
+	/**
+	 * Used to capture repair data from changes within a transaction. The repair data will be
+	 * add to the same {@link RepairDataStore} until {@link UndoRedoManager.trackCommit} is called.
+	 * @param startRevision - the revision of the first commit in the transaction
+	 */
+	public trackRepairData(change: Delta.Root, startRevision: RevisionTag) {
+		if (this.pendingRepairData === undefined) {
+			this.pendingRepairData = this.repairDataStoryFactory();
+		}
+
+		this.pendingRepairData.capture(change, startRevision);
+	}
+
+	/**
+	 * Adds the provided commit to the undo commit tree. Also deals with commits created by calling {@link UndoRedoManager.undo}.
+	 */
+	public trackCommit(commit: GraphCommit<TChange>) {
+		if (commit.type === GraphCommitType.Undo) {
+			// Currently no need to handle undo commits
+			return;
+		}
+
+		const parent = this.headUndoCommit;
+		const repairData = this.pendingRepairData ?? this.repairDataStoryFactory();
```

Comment:
This is not quite right: the repair data for a commit is different from the repair data for the changes that make up the commit (unless there's only one change in the commit).
I don't think this class should have to care at all about the repair data for the individal changes that make up a transaction. I think this means `trackRepairData` and `pendingRepairData` can go away. Happy to chat to clarify more.

Reply:
This also means that the old code for traccking the repair data of transactions can remain as it was.

Reply:
Ah I see, that makes sense. This does mean that getting the repair data for a commit will be a bit more complicated though since the forest is updated per change rather than per commit.

Reply:
Good point. You'll need to clone the forest when a transaction starts so you can apply the whole commit to it.